### DISCUSSION TITLE
Topology optimization

### DISCUSSION THREAD

**Discussion by skbizport** [seamplex/feenox#14] [General]
Hi

Appreciate great work on this library. 

Is there any example of how to use feenox to put together implementation of topology optimization.

Thx
Shailesh

**Comment by gtheler** [comment]
Hi. There's no example of this. I've tried to do something with a company from Israel a couple of years ago but work stalled.

There are two main things here:
 1. to have an external tool to perform the removal or addition of materials in the mesh file (this was more or less accomplished by this company: it would start with the original CAD, perform some operations and en up with an STL with the "new" geometry for the next step)
 2. to be able to recover the faces with boundary conditions after each step of the removal/addition of elements (since the geometry was in STL, we could use Gmsh to re-mesh by re-parametrizing but identifying the triangles belonging to the original faces with BCs was tricky).

If you have any cool ideas for each of these two issues, let's discuss it here.

**Reply by skbizport** [reply]
Not an expert on this - seems like for #2, may be first rearrange the mesh so that all BC faces appear before all others. when mesh is modified (material removed), hopefully those BC triangles will still remain in place with same indexing. 

There is another project on github which uses PETScL https://github.com/thsmit/TopOpt_in_PETSc_wrapped_in_Python. May be adapt that?

Thx

**Comment by gtheler** [comment]
If the elements are directly removed from the original mesh, that wouldn't be a problem. But the workflow I stated does not use the FEM mesh, it uses a different triangulation to remove volume from the original CAD independently of the Gmsh mesh that Feenox uses to solve the mechanical problem.

Honestly I don't know that project you mention. Maybe I'm wrong but if that uses PETSc directly then you wouldn't be needing FeenoX.

The idea is akin to the case of optimizing the length of the tuning fork to get a desired base frequency: https://www.seamplex.com/feenox/examples/modal.html#optimizing-the-length-of-a-tuning-fork
What you need is an "outer" iterator that for each steps gives FeenoX a proper mesh and reads back one or more figures of merit. According to the results of each step, this outer iterator would carve the CAD, mesh it and create a new `.msh` file for the next step. Does that make sense?
